DependencyInjectionContainer/DIContainer.cs

```diff
@@ -1,130 +1,265 @@
+            IEnumerable<TTypeToResolve> ResolveLocal() =>
+                registeredServices
+                    .Where(service => service.Key == typeof(TTypeToResolve))
+                    .Select(service => (TTypeToResolve)GetOrCreateServiceImplementation_SaveIfSingleton(service, ResolveSource.Local));
+
+            IEnumerable<TTypeToResolve> ResolveNonLocal()
+                => ContainerParent != null ? ContainerParent.ResolveMany<TTypeToResolve>() : Enumerable.Empty<TTypeToResolve>();
+
+            IEnumerable<TTypeToResolve> ResolveNonLocalImplementationTypesWhichWereNotResolved(IEnumerable<TTypeToResolve> resolvedServices) =>
+                ResolveNonLocal().Where(resolved => !resolvedServices.Any(item => item.GetType() == resolved.GetType()));
+        }
+
+        public void Dispose()
+        {
+            if (!IsDisposed)
             {
-                return serviceToResolve.Implementation;
+                servicesInstanceList.Dispose();
+                servicesInstanceList = null;
+                ContainerParent = null;
+                GC.SuppressFinalize(this);
+                IsDisposed = true;
             }
+        }
 
-            var implementation = GetCreatedImplementationForService(serviceToResolve);
+        private void ThrowIfDisposed()
+        { 
+            if (IsDisposed)
+            {
+                throw new NullReferenceException("This container was disposed");
+            }
+        }
 
-            if(serviceToResolve.Lifetime == ServiceLifetime.Singleton)
+        private object Resolve(Type typeToResolve, ResolveSource resolveSource)
+        {
+            if (typeToResolve.IsValueType)
             {
-                serviceToResolve.Implementation = implementation;
+                throw new ArgumentException("Can resolve only reference types");
             }
 
-            return implementation;
+            if (typeToResolve.IsEnumerable())
+            {
+                typeToResolve = typeToResolve.GetGenericArguments()[0];
+                return InvokeGenericResolveMany(typeToResolve, this, resolveSource);
+            }
 
-            object GetCreatedImplementationForService(Service service)
+            switch (resolveSource)
             {
-                var constructorInfo = service.ImplementationType.GetConstructors().First();
+                case ResolveSource.NonLocal:
+                    if (ContainerParent == null)
+                    {
+                        throw new NullReferenceException("Current container do not have parent");
+                    }
+                    return ContainerParent.Resolve(typeToResolve, ResolveSource.Any);
 
-                var parameters = constructorInfo.GetParameters().Select(parameter => Resolve(parameter.ParameterType)).ToArray();
+                case ResolveSource.Any:
+                    Service foundService;
+                    if (TryGetRegistration(typeToResolve, ResolveSource.Any, out foundService))
```

Comment:
You can inline out param
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/out-parameter-modifier#calling-a-method-with-an-out-argument